ouroboros-consensus/src/Ouroboros/Consensus/Node/Run/Abstract.hs

```diff
@@ -59,8 +60,8 @@ class (ProtocolLedgerView blk, ApplyTx blk) => RunNode blk where
   nodeEncodeApplyTxError :: Proxy blk -> ApplyTxErr blk -> Encoding
 
   -- Decoders
-  nodeDecodeHeader       :: forall s. NodeConfig (BlockProtocol blk) -> Decoder s (Header blk)
-  nodeDecodeBlock        :: forall s. NodeConfig (BlockProtocol blk) -> Decoder s blk
+  nodeDecodeHeader       :: forall s. NodeConfig (BlockProtocol blk) -> Decoder s (Lazy.ByteString -> Header blk)
+  nodeDecodeBlock        :: forall s. NodeConfig (BlockProtocol blk) -> Decoder s (Lazy.ByteString -> blk)
```

Comment:
Not happy about these types but I'm not sure it would be any better removing the `Decoder` and using `Lazy.ByteString -> Either DecoderError blk` instead

Reply:
Indeed that would not be a drop in replacement, since the storage system needs the decoder to find where the record boundaries are.